### DISCUSSION TITLE
Host for MSC on renesas RA

### DISCUSSION THREAD

**Discussion by abakosh** [hathach/tinyusb#2062] [General]
I am using tinyusb host functionality on the Renesas RA family. I am able to mount or unmount USB sticks successfully. However, for some devices, the stack gets stuck and I keep getting the following message when I enabled the debug:
[0:] USBH Defer Attach until the current enumeration
I follow it to this part of the code
`      
case HCD_EVENT_DEVICE_ATTACH:
        // due to the shared _usbh_ctrl_buf, we must complete enumerating
        // one device before enumerating another one.
        if ( _dev0.enumerating )
        {
          TU_LOG_USBH("[%u:] USBH Defer Attach until current enumeration complete\r\n", event.rhport);
          osal_queue_send(_usbh_q, &event, in_isr);
        }
`
Do you know what could be the reason for this behavior, is there anyone got the same issue?

I appreciate your help

**Comment by mcuenthusiast** [comment]
I also faced similar issue sometimes, but it got resolved automatically some times I haven't look for RCA yet.

Are you using freeRTOS with IAR?

**Reply by abakosh** [reply]
No I am not using RTOS.

I am still trying to figure out the root cause. Nothing so far

**Comment by mcuenthusiast** [comment]
I am using IAR and FreeRTOS with EK-RA6M5.
I am also facing this issue sometimes but when I Used e2studio with
EK-RA6M5 i Didn't get any issue it is working fine every.

On IAR i am facing some issues regarding interrupt.

In recent PR for USBHS(not merged) some modifications have been made for
attach interrupt. you can try that changes once.

Not sure if this is some compiler dependent or not.

Also try to plug out and plug in flash drive again.
I think that helps.

**Comment by abakosh** [comment]
When you are building with IAR, what are you doing regarding the __builtin_ctz function, it is not native to the IAR environment. 
I am using the following:
#define __builtin_ctz(x)             __iar_builtin_CLZ(__iar_builtin_RBIT(x))
It is declared implicitly if you don't implement it, how is it for you?

**Reply by mcuenthusiast** [reply]
I have implemented my own function to count number of trailing zeros, Also just for your information I have tested Host code and have done file operations on EK-RA6M5 it is working fine with e2-studio.

**Reply by abakosh** [reply]
I am having weird behavior if I stick the USB stick and remove it very fast, it is very easily producible. Did you see this behavior on e2studio, the tuh_task stuck forever after that

**Reply by mcuenthusiast** [reply]
hi @abakosh,

Did you find the root cause for the issue you had earlier which you mentioned in the thread? - HCD_EVENT_DEVICE_ATTACH - event calling again and again.? which board are you using?

Also Did You face below issue ever?

1. in debug mode code gets stuck in below line, due to breakpoint.
// Attempt to transfer on a busy endpoint, sound like an race condition !
  TU_ASSERT(ep_state->busy == 0);

**Comment by abakosh** [comment]
@perigoso Hello, thank you for your work on the RA support. Did you have the chance to test the host code with the IAR compiler?

**Reply by perigoso** [reply]
Hi,
unfortunately i never got to thoroughly test the host side of things, my target application was a device and that's where my focus was, and I never worked with IAR, to add insult to injury

**Comment by mcuenthusiast** [comment]
Hi @abakosh ,

Have you faced any issue regarding NRDY received, after sending MSC_STAGE_CMD command? I am getting NRDY for pipe4, have you made any changes regarding to endian ness or any thing related to compiler specific with IAR.- I already made the changes you mentioned in PR - https://github.com/hathach/tinyusb/pull/2063.
